overlord/devicemgmtstate/devicemgmtmgr.go

```diff
@@ -0,0 +1,546 @@
+	"github.com/snapcore/snapd/features"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/overlord/configstate/config"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/overlord/swfeats"
+	"github.com/snapcore/snapd/store"
+	"gopkg.in/tomb.v2"
+)
+
+const (
+	deviceMgmtStateKey = "device-mgmt"
+
+	defaultExchangeLimit    = 10
+	defaultExchangeInterval = 5 * time.Minute
+
+	awaitSubsystemRetryInterval = 30 * time.Second
+)
+
+var (
+	timeNow = time.Now
+
+	deviceMgmtCycleChangeKind = swfeats.RegisterChangeKind("device-management-cycle")
+)
+
+// MessageHandler processes request-message messages of a specific kind.
+// Caller must hold state lock when using this interface.
+type MessageHandler interface {
+	// Validate checks subsystem-specific constraints (authorization, payload schema, etc).
+	// Returns nil if valid, error describing the validation failure otherwise.
+	Validate(st *state.State, msg *PendingMessage) error
+
+	// Apply processes a request-message and returns a change ID.
+	Apply(st *state.State, reqAs *PendingMessage) (changeID string, err error)
+
+	// BuildResponse converts a completed change into a response body and status.
+	BuildResponse(chg *state.Change) (body map[string]any, status asserts.MessageStatus)
+}
+
+// ResponseMessageSigner can sign response-message assertions.
+type ResponseMessageSigner interface {
+	SignResponseMessage(accountID, messageID string, status asserts.MessageStatus, body []byte) (*asserts.ResponseMessage, error)
+}
+
+// PendingMessage represents a request-message being processed.
+// Messages remain pending until their associated change completes,
+// at which point a response is queued and the message is removed.
+type PendingMessage struct {
+	BaseID      string    `json:"base-id"`
+	SeqNum      int       `json:"seq-num"`
+	Kind        string    `json:"kind"`
+	AccountID   string    `json:"account-id"`
+	AuthorityID string    `json:"authority-id"`
+	Source      string    `json:"source"`
+	Received    time.Time `json:"received"`
+
+	Devices    []string  `json:"devices"`
+	ValidSince time.Time `json:"valid-since"`
+	ValidUntil time.Time `json:"valid-until"`
+
+	Body string `json:"body"`
+
+	ChangeID        string `json:"change-id,omitempty"`        // Set when subsystem change is created
+	ValidationError string `json:"validation-error,omitempty"` // Set when validation fails
+	ApplyError      string `json:"apply-error,omitempty"`      // Set when apply fails
+}
+
+func (msg *PendingMessage) ID() string {
+	if msg.SeqNum != 0 {
+		return fmt.Sprintf("%s-%d", msg.BaseID, msg.SeqNum)
+	}
+
+	return msg.BaseID
+}
+
+// exchangeConfig holds parameters for the message exchange task.
+type exchangeConfig struct {
+	// Limit is the maximum number of request messages to fetch.
+	Limit int
+}
+
+// DeviceMgmtState holds the persistent state for device management operations.
+type DeviceMgmtState struct {
```

Comment:
not sure this should be an exported struct, we can create a exported equivalent type in export_test.go if needed